code/HabitTracker/app/src/main/java/com/example/habittracker/DatabaseManager.java

```diff
@@ -240,4 +251,162 @@ public void onFailure(@NonNull Exception e) {
+        DocumentReference doc = usersColRef
+                .document(username);
+        doc.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
+                if (task.isSuccessful()) {
+                    User user;
+                    DocumentSnapshot document = task.getResult();
+                    if(!document.exists()){
+                        callback.onCallbackFailed();
+                        return;
+                    }
+                    String hashedPassword = (String) document.getData().get("hashedPassword");
+                    user = new User(
+                            username,
+                            hashedPassword
+                    );
+                    callback.onCallbackSuccess(user);
+                }
+                else{
+                    callback.onCallbackFailed();
+                }
+            }
+        });
+    }
+
+    public void userExists(String username, BooleanCallback callback){
+        // Users -> userid (key) -> Habits -> habitTitle (key) -> HabitEvents
+        DocumentReference doc = usersColRef
+                .document(username);
+        doc.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
+                if (task.isSuccessful()) {
+                    boolean flag = false;
+                    DocumentSnapshot document = task.getResult();
+                    if(document.exists()){
+                        flag = true;
+                    }
+                    callback.onCallbackSuccess(flag);
+                }
+                else{
+                    callback.onCallbackFailed();
+                }
+            }
+        });
+    }
+
+    public void checkPassword(String username, String hashedPassword, BooleanCallback callback){
+        // Users -> userid (key) -> Habits -> habitTitle (key) -> HabitEvents
+        DocumentReference doc = usersColRef
+                .document(username);
+        doc.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
+                if (task.isSuccessful()) {
+                    boolean flag = false;
+                    DocumentSnapshot document = task.getResult();
+                    if(!document.exists()){
+                        callback.onCallbackFailed();
+                        return;
+                    }else if(hashedPassword.equals((String)document.getData().get("hashedPassword"))){
+                        flag = true;
+                    }
+                    callback.onCallbackSuccess(flag);
+                }
+                else{
+                    callback.onCallbackFailed();
```

Comment:
Same comment here -
I was hoping to see a UserDetailsCallback with methods onCallbackSuccess(String username, String hashedPassword).
This way, the controller class like the LoginActivity can create the User object inside its implementation of UserDetailsCallback.onCallbackSuccess and start a new Activity from within the callback.
It might be harder for the controller class to take actions inside its implementation of UserDetailsCallback.onCallbackSuccess while accessing variables that lie outside of the callback function. If I recall correctly, if you want use an outer class variable inside an anonymous inner class (which is the callback implementation), the outer class variables need to be declared final.